lgc/patch/PatchInvariantLoads.cpp

```diff
@@ -138,7 +139,22 @@ bool PatchInvariantLoads::runImpl(Function &function, PipelineState *pipelineSta
 
   auto &options = pipelineState->getShaderOptions(shaderStage);
   bool clearInvariants = options.disableInvariantLoads;
-  bool aggressiveInvariants = options.aggressiveInvariantLoads && !clearInvariants;
+  bool aggressiveInvariants = !clearInvariants && options.aggressiveInvariantLoads == Enable;
+
+  if (!clearInvariants && options.aggressiveInvariantLoads == Auto &&
+      pipelineState->getTargetInfo().getGfxIpVersion().major >= 10) {
+    switch (function.getCallingConv()) {
```

Comment:
Looking at the code where the Auto option enables 'aggressiveInvariants' for gfx10+ and for some shader types, do I understand correctly that your experiments have shown:
+ Auto is not convincingly profitable for PS/CS on any asics you tested
+ Auto is not convincingly profitable for VS-like on gfx9

Reply:
From my experiments:
- CS yields little as these shaders commonly write memory so invariant optimization cannot normally be applied, so avoid compile time cost of iterating these shaders.
- PS has large benefits in many cases, but also many cases with large negative impact, so a fine grain heuristic is required to Auto these.  I have investigated heuristics, but nothing has provided satisfactory results yet.

I have not tested GFX9 due to not having the hardware setup currently, but it is possible VS-like may provide similar benefit on GFX9.  My expectation is to follow up after GFX9 testing.

Reply:
Sounds good, thanks for explaining.